documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,555 @@
+* *Pipe*: The duplex communication construct this protocol is communicated over.  This is a Unix Domain Socket on *nix systems and a Named Pipe on Windows.
+* *IPC Message*: The base unit of communication over the Diagnostic IPC Protocol. Is made up of a Header and a Payload.
+* *Header*: A struct containing version and metadata.
+* *Payload*: An opaque chunk of data that is Command specific.
+* *Optional Continuation*: The reuse of the pipe for application specific communication. This communication does not need to adhere to any requirements listed in this spec, e.g., this could be a stream of custom encoded data that is Command specific.
+
+### General Flow
+
+All communication with the Diagnostic Server will begin with a Diagnostic IPC Message sent from the client to the server.  The server will respond with a Diagnostic IPC Message.  After this, the client and runtime _may_ reuse the Pipe for any Command specific communication which is referred to as an Optional Continuation.
+
+![Generic Flow](ipc-protocol-genericflow.svg)
+
+```
+runtime <- client : [ Header ][ Payload ]
+runtime -> client : [ Header ][ Payload ]
+optional:
+    runtime <-> client : [ Optional Continuation ]
+connection closed
+```
+
+Example flow for EventPipe:
+```
+runtime <- client : [ ver; size; EventPipe CollectTracing ][ stream config struct  ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; EventPipe TracingStarted ][ stream started struct ] <- Diagnostic IPC Message
+runtime -> client : [ stream of netperf data ]                                       <- Optional Continuation
+
+// stop message is sent on another connection
+
+connection closed
+```
+
+Example flow for hypothetical application with custom communication protocol:
+```
+runtime <- client : [ ver; size; start_app_session_command    ][ app config struct      ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; app_session_started_response ][ session started struct ] <- Diagnostic IPC Message
+
+Optional Continuation using app specific protocol
+
+runtime <- client : [ command specific protocol ]                                         <- Optional Continuation
+runtime -> client : [ command specific protocol ]                                         <- Optional Continuation
+Eventually...
+runtime <- client : [ command specific protocol (close connection) ]                      <- Optional Continuation
+connection closed
+```
+
+### Transport
+
+The protocol will be communicated over a platform-specific transport.  On Unix/Linux based platforms, a Unix Domain Socket will be used, and on Windows, a Named Pipe will be used.
+
+#### Naming and Location Conventions
+
+Unix Domain Sockets (MacOS and *nix):
+
+The location of the socket is determined by a call to `GetTempPathA()` in the PAL.  This has a fallthrough behavior and attempts to use the following three locations, in order:
+1. (Mac only) App Group Container if `AppGroupId` is not null
```

Comment:
If we could write all the sockets in a global location that significantly eases the ability to enumerate them. I assume to do an enumeration now I'd first need to enumerate all ApplicationGroups, then enumerate sockets at each of those directories. However I don't know if we'd have permission to do this from within an AppGroup process?

Reply:
This reflects the current implementation, but I agree.  We can define what we want in the spec, and file an issue to track cases of misalignment.  In this case, we do get free, guaranteed cleanup if we use the unix tmp directories (even on Mac).  I don't think our diagnostic tools even look outside the temp directory as it is anyway, so it might be good to do the work to ensure we are always in the `tmp` dir, and potentially creating out own parent directory that sets the perms as we want them.